wordpress-mobile/WordPress-iOS
# Issues/7430 missing posts

Refs #7430 
This has been a tricky glitch to sort out. This time round, while reviewing the code it occurred to me that if a site owner were to change the publish date of a previously published post it would be possible to recreate the behavior reported by our users due to the way overlaps were detected.  

When an overlap is detected the app assumes that all available content is present in the app and no gap marker is inserted to indicate there might be missing posts. 

Here's the scenario:
- A user follows the Coffee tag and syncs its posts, including one from Juan's Coffee Blog.
- Later, Juan publishes some more posts, but decides to change the date of an earlier post (one that was synced by our user) so it shows up at the top of his site.  (Poor Juan hasn't heard of sticky posts.)
- Our faithful users returns to the app some time later and syncs the Coffee tag again. 
- The newly fetched posts include the one with the changed date. 
- When the app checks to see if any of the newly fetched posts have are already synced, its going to flag the one with the date changed.   We actually don't want this. Since the post is changing its position in the list we want to treat it like new content, not an overlap. 

This PR improves the way overlaps are detected to check that dates match between what was synced and what was cached. If dates match then there is an overlap.  If dates do not match we do not treat it as an overlap. 

This should fix the problem our users are reporting.  (Unless there's something else going on.)

To test:
Testing is a little tricky. 
1. Edit ReaderPostService and set `ReaderPostServiceNumberToSync` to a low number like `5`. 
2. Edit ReaderStreamViewController and set `refreshInterval` to something like `10`. 
3. Comment out `syncHelper.syncMoreContent()` on line 1402 of ReaderStreamViewController to disable the loadMore mechanism otherwise it will conflict with a blog with only a handful of posts.
These will make it easier to test. 
4. Create a test blog and follow it. 
5. Add 5 test posts via a desktop browser
6. In the app go to Reader > Following > Manage > Your site and view its posts.
7. Tap the back button to leave the stream. 
8. Return to the blog and create ten more test posts.  Basically you want enough to create a gap
9. Edit one of the first posts you published so its publish date is the most recently published post.
10. Go back to the app and tap to view the posts for the site again. 
11. Confirm you see a gap marker. 

This seems to work for testing. Something odd I noticed was one of the posts was missing from the results, but it eventually showed up.  I'm tempted to chalk this up to caching but I'd be curious to see if it turns up in other tests. 

Needs Review: @frosty could I trouble you with this one?
